Branden Connolly <br/>
Assignment 11 <br/>
11/21/2016 <br/>

#Problem 1(a).  Reading Amazon Reviews.

In this problem, we will analyze Amazon reviews to determine what characteristics make them most helpful.

Download the file of Amazon gourmet food reviews from the [Stanford Large Network Dataset Collection](https://snap.stanford.edu/data/web-FineFoods.html).   (Your computer may already have a utility installed that can unzip the archive as a text file; if not, [7-zip](http://www.7-zip.org/) is a useful utility for Windows. You can also do a web search for ```open .gz files online``` and use an online utility.)

Create a pandas DataFrame object with the following entries for each review:

* Product ID
* Number of people who voted this review helpful
* Total number of people who rated this review
* Rating of the product
* Text of the review

For the second and third of these, the information will be given in the file as ```1/5```, which would correspond to 1 vote for helpful out of 5 people who rated the review.


In [5]:
import pandas as pd
import numpy as np


with open("/Users/brandenconnolly/Documents/DS710/ds710fall2016assignment11/finefoods.txt",encoding = "ISO-8859-1") as food_rev:

    dict_list = []
    column_names = ('Product ID', 'Review Helpfulness', 'Total Ratings', 'Rating', 'Review')
    

    while True:
        line = food_rev.readline()
        
        if not line:
            break
        
        line=line.strip()
        line_split = line.split(":")
        
        if(line_split[0] == "product/productId"):
            prod = line_split[1].strip()
        elif(line_split[0] == "review/userId"):
            pass
        elif(line_split[0] == "review/profileName"):
            pass
        elif(line_split[0] == "review/helpfulness"):
            rating_split = line_split[1].split("/")
            review_help = rating_split[0]
            tot_rating = rating_split[1]
        elif(line_split[0] == "review/score"):
            rating = line_split[1]
        elif(line_split[0] == "review/time"):
            pass
        elif(line_split[0] == "review/summary"):
            pass
        elif(line_split[0] == "review/text"):
            review = line_split[1]
        elif(line_split[0] == ""):
            dict_list.append(dict(zip(column_names, [prod,review_help,tot_rating,rating,review])))
        else:
            pass

        
finefood_df = pd.DataFrame(dict_list)

print(finefood_df.head(5))

   Product ID Rating                                             Review  \
0  B001E4KFG0    5.0   I have bought several of the Vitality canned ...   
1  B00813GRG4    1.0   Product arrived labeled as Jumbo Salted Peanu...   
2  B000LQOCH0    4.0   This is a confection that has been around a f...   
3  B000UA0QIQ    2.0   If you are looking for the secret ingredient ...   
4  B006K2ZZ7K    5.0   Great taffy at a great price.  There was a wi...   

  Review Helpfulness Total Ratings  
0                  1             1  
1                  0             0  
2                  1             1  
3                  3             3  
4                  0             0  


#Problem 1(b).  Analyzing review text.

Add columns to your DataFrame for the length of a review, the number of exclamation points in a review, and the fraction of people who rated a review helpful. You should calculate the fraction who rated a review helpful using the two columns you made in 1a, and a ratio of 1 helpful rating out of 5 total ratings should be entered as 0.2, not the string ```1/5```. If no people voted on whether a problem was helpful, the corresponding entry in your percentage helpful column should be ```NaN```.

In [14]:
#length of review
finefood_df["Len_of_Review"] = finefood_df.Review.str.len()
#number of exclaimation points
finefood_df["Exclaimation_Points"] = finefood_df.Review.str.count('!')
#helpful_review 
#finefood_df["Helpful_Review"] = pd.to_numeric(finefood_df.Helpfulness).div(pd.to_numeric(finefood_df["Total Ratings"]))
#Need to seperate

finefood_df["Helpfulness_NUM"] = pd.to_numeric(finefood_df["Review Helpfulness"])
finefood_df["Total_NUM"] = pd.to_numeric(finefood_df["Total Ratings"])

finefood_df["Helpful_Review"] = finefood_df.Helpfulness_NUM.div(finefood_df.Total_NUM)

finefood_df.head(5)

Product ID Rating                                             Review  \
0  B001E4KFG0    5.0   I have bought several of the Vitality canned ...   
1  B00813GRG4    1.0   Product arrived labeled as Jumbo Salted Peanu...   
2  B000LQOCH0    4.0   This is a confection that has been around a f...   
3  B000UA0QIQ    2.0   If you are looking for the secret ingredient ...   
4  B006K2ZZ7K    5.0   Great taffy at a great price.  There was a wi...   

  Review Helpfulness Total Ratings  Len_of_Review  Exclaimation_Points  \
0                  1             1            264                    0   
1                  0             0            191                    0   
2                  1             1            510                    0   
3                  3             3            220                    0   
4                  0             0            141                    0   

   Helpfulness_NUM  Total_NUM  Helpful_Review  
0                1          1             1.0  
1                0          0             NaN  
2                1          1             1.0  
3                3          3             1.0  
4                0          0             NaN

#Problem 1(c).  Summary statistics.

How many reviews are in the data set?  What is the average length of a review (in characters)?  What is the average rating?  What is the greatest number of exclamation marks used in a single review?  Use the pandas package to answer these questions, then summarize your results in a markdown cell.

In [15]:
finefood_df.describe()

/Users/brandenconnolly/anaconda/lib/python3.5/site-packages/numpy/lib/function_base.py:3834: RuntimeWarning: Invalid value encountered in percentile
  RuntimeWarning)


Len_of_Review  Exclaimation_Points  Helpfulness_NUM     Total_NUM  \
count  568454.000000        568454.000000    568454.000000  568454.00000   
mean      389.031214             0.702817         1.743817       2.22881   
std       353.763842             1.478514         7.636513       8.28974   
min         1.000000             0.000000         0.000000       0.00000   
25%       171.000000             0.000000         0.000000       0.00000   
50%       283.000000             0.000000         0.000000       1.00000   
75%       484.000000             1.000000         2.000000       2.00000   
max     11899.000000            84.000000       866.000000     923.00000   

       Helpful_Review  
count   298402.000000  
mean         0.776975  
std          0.346321  
min          0.000000  
25%               NaN  
50%               NaN  
75%               NaN  
max          3.000000

There are 568,454 reviews in the dataset <br /> 
The average length of a review is 389 words (rounded to zero decimals)<br /> 
The most number of exlaimation points is 84 in one review <br /> 




#Problem 1(d).  Export.

Save your DataFrame as a .csv file suitable for future analysis in R.  Your .csv file should not include the review text column, as the presence of commas and quotation marks will make reading the file difficult.  You should also convert entries from ```NaN``` to the empty string before saving.

In [18]:
finefood_df_op = finefood_df.drop('Review',1)
finefood_df_op.fillna("",inplace = True)
finefood_df_op.to_csv("/Users/brandenconnolly/Documents/DS710/ds710fall2016assignment11/finefoods_OP.txt",index = False)